src/modeltools/CategoryTheory.jl

```diff
@@ -71,81 +67,58 @@ function ⊔(f::FinSetMorph, g::FinSetMorph)
-function (f::FinSetMorph)(g::G) where G <: AbstractGraph
-    dom(f) == vertices(g) || throw(DomainError(vertices(g), "dom(f) = $(dom(f)) but nv(g) = $(nv(g))"))
-    ϕ = func(f)
-    map(edges(g)) do e
-        s,t = e.src, e.dst
-        Edge(ϕ(s), ϕ(t))
-    end |> Graph
-end
-
-# +
-"""    GraphMorph{T,F} <: Morph
-
-morphisms in the category of Finite Graphs. The objects must be a subtype of AbstractGraph.
-
-You can take a `FinSetMorph` and lift it to a graph homomorphism. This is the functor that
-takes the finite set `1:n`, to the empty graph with `n` vertices.
+"""    Decorated{M,T}
 
+a decoration applied to the objects of a morphism, where M is a type of morphism and
+type T is the category of the decoration
 """
-struct GraphMorph{T, F} <: AbstractMorph
-    dom::T
-    codom::T
-    fun::F
+mutable struct Decorated{M}
+    f::M
+    d::AbstractArray{AbstractModel}
 end
 
-"""    GraphMorph(g::AbstractGraph, f::FinSetMorph)
+# Handle creating a decorated morphism with an array of a single type
+Decorated(f, d::AbstractArray{T}) where T<:AbstractModel = Decorated(f, Vector{AbstractModel}(d))
+# Handle creating a decorated morphism from a single decoration
+Decorated(f, d::T) where T<:AbstractModel = Decorated(f, Vector{AbstractModel}([d]))
 
-is defined to be the graph homomorphism you get by functorially lifting `f`.
-That is, `f` acts on the vertex set of `g` as an `Int->Int` function, and then
-must act on the edges consistently.
-"""
-GraphMorph(g::AbstractGraph, f::FinSetMorph) = GraphMorph(g, f(g), f)
+# Get the domain or codomain of a decorated morphism
+dom(m::Decorated) = dom(m.f)
+codom(m::Decorated) = codom(m.f)
 
-dom(m::GraphMorph) = m.dom
-codom(m::GraphMorph) = m.codom
-func(m::GraphMorph) = begin
-    f = func(m.fun)
-    return i->f(i)
+# Get the decorations of a decorated morphism
+function decorations(m::Decorated)
+  return m.d
+end
+# Get the decorations of AbstractModel T of a decorated morphism
+function decorations(m::Decorated, ::Type{T}) where T<:AbstractModel
+  filter(x -> isa(x,T), decorations(m))
 end
 
-"""    verify(m::GraphMorph)
+# Remove the decoration of a decorated morphism, and return the original morphism
+function undecorate(m::Decorated)
+  return m.f
+end
 
-validate a graph homomorphism by checking that all the edges in `dom(m)` and map to edges in `codom(m)`.
-"""
-verify(m::GraphMorph) = begin
-    dom(m.fun) == vertices(dom(m)) || return false
-    codom(m.fun) == vertices(codom(m)) || return false
-    E = Set(edges(codom(m)))
-    f = func(m)
-    map(edges(dom(m))) do e
-        u,v = f(e.src), f(e.dst)
-        if u > v
-            u,v = v,u
-        end
-        Edge(u,v) in E
-    end |> all
+# Add a decoration to a decorated morphism
+function add_decoration!(m::Decorated, decoration::AbstractModel)
+  push!(decorations(m), decoration)
 end
 
-# +
-"""    Decorated{M,T}
+# Add a collection of decorations to a decorated morphism
+function add_decorations!(m::Decorated, decorations)
```

Comment:
I think this should be `Base.append!(m::Decorated, decorations) = append!(m.d, decorations)` for consistency with the standard library naming convention?

Reply:
done.